GH952: Refactor DotNetCore args string->ArgumentBuilder

devlead opened the pull request:
This is a proposed fix for #952

Currently the DotNetCore DotNetCoreExecute and DotNetCoreRun uses a quoted single string for args, this breaks in some scenarios also wont allow filtering sensitive data.

This PR
* Refactors arguments from string to ProcessArgumentBuilder
  * This takes benefit from all ProcessArgumentBuilder extensions available
  * Enables RenderSafe to work for verbose logging
* Adds `CopyTo(ProcessArgumentBuilder target)` method to `ProcessArgumentBuilder`
  * This allows without exposing internals to append the content of one `ProcessArgumentBuilder` to another
* Adds an `IsEmpty()` method to `ProcessArgumentBuilder`
  * This indicates if `ProcessArgumentBuilder` has any args / would render empty.

patriksvensson commented on src/Cake.Core/IO/ProcessArgumentBuilder.cs:
```diff
@@ -83,6 +84,33 @@ public string FilterUnsafe(string source)
+        /// Copies all the arhuments of the current <see cref="ProcessArgumentBuilder"/> to target <see cref="ProcessArgumentBuilder"/>.
+        /// </summary>
+        /// <param name="target">The copy target.</param>
+        public void CopyTo(ProcessArgumentBuilder target)
```
Maybe it makes more sense exposing the tokens as a `IReadOnlyList<IProcessArgument>` and add those two methods as extension methods since they only require read access.

devlead replied:
Couldn't that potentially leak sensitive tokens?

patriksvensson replied:
Well, the person who got access to the token could just as well use the `Render` method, or am I missing something?

devlead commented:
@patriksvensson refactored CopyTo / IsNullOrEmpty to pure extension methods, so think I've addressed your feedback. Also rebased against latest develop.

patriksvensson commented on src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs:
```diff
@@ -380,5 +380,39 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
         }
+
+        /// <summary>
+        /// Indicates whether a <see cref="ProcessArgumentBuilder"/> is null or renders empty.
```
Should be `indicates`.

devlead replied:
Unsure what you mean :)
![image](https://cloud.githubusercontent.com/assets/1647294/15837190/80f9d504-2c38-11e6-9641-a1783ecc2b82.png)

patriksvensson replied:
Did I just have a stroke? I could have sworn it said `indiates` when I read it.

patriksvensson commented on src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs:
```diff
@@ -380,5 +380,39 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
+        }
+
+        /// <summary>
+        /// Copies all the arhuments of the source <see cref="ProcessArgumentBuilder"/> to target <see cref="ProcessArgumentBuilder"/>.
```
Should be `arguments`.

patriksvensson commented on src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs:
```diff
@@ -380,5 +380,39 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
+        /// <summary>
+        /// Copies all the arhuments of the source <see cref="ProcessArgumentBuilder"/> to target <see cref="ProcessArgumentBuilder"/>.
+        /// </summary>
+        /// <param name="source">The copy source.</param>
```
Could probably be something like `The process argument builder to copy from`.

patriksvensson commented on src/Cake.Core/Extensions/ProcessArgumentListExtensions.cs:
```diff
@@ -380,5 +380,39 @@ public static ProcessArgumentBuilder AppendQuotedSecret(this ProcessArgumentBuil
+        /// Copies all the arhuments of the source <see cref="ProcessArgumentBuilder"/> to target <see cref="ProcessArgumentBuilder"/>.
+        /// </summary>
+        /// <param name="source">The copy source.</param>
+        /// <param name="target">The copy target.</param>
```
Could probably be something like `The process argument builder to copy to.`.

patriksvensson commented on src/Cake.Core/IO/ProcessArgumentBuilder.cs:
```diff
@@ -103,5 +105,31 @@ public static ProcessArgumentBuilder FromString(string value)
+        /// <summary>
+        /// Gets the number of arguments contained in the <see cref="ProcessArgumentBuilder"/>.
+        /// </summary>
+        public int Count
```
Move above constructor.

patriksvensson commented on src/Cake.Core/IO/ProcessArgumentBuilder.cs:
```diff
@@ -103,5 +105,31 @@ public static ProcessArgumentBuilder FromString(string value)
+        /// Returns an enumerator that iterates thru the collection.
+        /// </summary>
+        /// <returns><see cref="IEnumerator"/></returns>
+        IEnumerator IEnumerable.GetEnumerator()
```
No comment necessary for explicit implementation.

patriksvensson commented on src/Cake.Core/IO/ProcessArgumentBuilder.cs:
```diff
@@ -103,5 +105,31 @@ public static ProcessArgumentBuilder FromString(string value)
+        /// Returns an typed enumerator that iterates thru the collection.
+        /// </summary>
+        /// <returns><see cref="IEnumerator{IProcessArgument}"/></returns>
+        IEnumerator<IProcessArgument> IEnumerable<IProcessArgument>.GetEnumerator()
```
Should be `Returns an enumerator that iterates through the collection.`

patriksvensson commented:
@devlead Left some minor comments. Looks good to me otherwise.

devlead commented:
@patriksvensson think I've addressed your feedback now.

patriksvensson commented:
Merged. Awesome work :+1:
